### PR TITLE
Add reservoir sampling option to writers

### PR DESCRIPTION
**Pull request by lpawluczuk**
This is implementation of reservoir sampling for stream of items on writers level. 

It addresses this issue: https://github.com/scrapinghub/exporters/issues/317

### CODE REVIEW

File: exporters/module_loader.py [hunk 1]
```diff
@@ -48,14 +48,22 @@ def load_stats_manager(self, options, metadata, **kwargs):
         from exporters.stats_managers.base_stats_manager import BaseStatsManager
         return self._load_module(options, metadata, BaseStatsManager, **kwargs)
 
+    def load_write_buffer(self, module_name, **kwargs):
+        from exporters.write_buffers.base_buffer import BaseBuffer
+        module_type = BaseBuffer
+        instance = self._instantiate_class(module_name, **kwargs)
+        if not isinstance(instance, module_type):
+            raise TypeError('Module must inherit from ' + str(module_type))
```
Comment (kalessin) [inline exporters/module_loader.py]: do we really need that any custom write buffer be a subclass of BaseBuffer? We don't do this type check in other classes loaded through options.

File: exporters/write_buffers/base_buffer.py [hunk 2]
```diff
@@ -118,7 +118,8 @@ class GroupingBufferFilesTracker(object):
     that is cleaned up when calling close().
     """
 
-    def __init__(self, formatter, compression_format):
+    def __init__(self, formatter=None, compression_format=None, *args, **kwargs):
```
Comment (kalessin) [inline exporters/write_buffers/base_buffer.py]: which is the point of having *args after two parameters with default values? In order args to actually have some value, you must always pass some value to formatter and to compression_format. So, either formatter and compression_format shouldn't have default values, or *args shouldn't be there.

File: exporters/write_buffers/base_buffer.py [hunk 3]
```diff
@@ -118,7 +118,8 @@ class GroupingBufferFilesTracker(object):
...
+        super(GroupingBufferFilesTracker, self).__init__(*args, **kwargs)
```
Comment (kalessin) [inline exporters/write_buffers/base_buffer.py]: super class here is object. Why need to call it?

File: exporters/write_buffers/base_buffer.py [hunk 4]
```diff
@@ -154,21 +155,35 @@ def get_current_buffer_file_for_group(self, key):
...
-class WriteBuffer(object):
+
+class BaseBuffer(object):
 
     def __init__(self, items_per_buffer_write, size_per_buffer_write,
-                 items_group_files_handler, compression_format='gz',
-                 hash_algorithm=None):
+                 formatter=None, compression_format='gz', hash_algorithm=None,
+                 items_group_files_handler=None, *args, **kwargs):
+        super(BaseBuffer, self).__init__(*args, **kwargs)
```
Comment (kalessin) [inline exporters/write_buffers/base_buffer.py]: also here, why to call super here? super class is object.

**Comment by kalessin** [comment]
I am having problems reviewing this PR. It has too many changes in a big commit and I don't really see the reason of many of the changes (not only the ones I commented, but also many of the rest). I think they are complicating unnecessarily the code. I was thinking on something much more simple. I think we don't need to change so many things to achieve the same.

If you create a new PR with only one commit which only adds the reservoir sampling class (and another that adds tests for the isolated class, it should be possible), I can do the changes needed in the exporters for fitting on it. It should be possible to do things much more simple.

**Comment by kalessin** [comment]
don't remove this PR, I will use it for reference

**Comment by kalessin** [comment]
I will proceed in the inverse way. I will provide the basis for adding custom write buffers and then you create a PR for the reservoir sampling.

File: exporters/write_buffers/base_buffer.py (hunk 3 above)
Comment (lpawluczuk) [inline exporters/write_buffers/base_buffer.py]: http://stackoverflow.com/questions/8611712/what-does-objects-init-do-in-python

"super() does not always return a proxy for the parent class. Instead, it returns a proxy for the next class in MRO. In single-inheritance there is no difference between MRO and the inheritance chain. In multiple-inheritance, MRO may result in a class on the other inheritance chain instead."

It was needed to make diamond inheritance work.

**Comment by kalessin** [comment]
This PR is old and has been replaced by other ones

**Pull request closed without merging**